src/chain_sync/tipset_syncer.rs

```diff
@@ -1094,20 +1083,25 @@ async fn sync_messages_check_state<DB: Blockstore + Send + Sync + 'static>(
         .try_for_each(|batch| async {
             for full_tipset in batch {
                 let current_epoch = full_tipset.epoch();
-                let timer = metrics::TIPSET_PROCESSING_TIME.start_timer();
-                validate_tipset(
-                    state_manager.clone(),
-                    &chainstore,
-                    bad_block_cache,
-                    full_tipset.clone(),
-                    genesis,
-                    invalid_block_strategy,
-                )
-                .await?;
-                drop(timer);
-                chainstore.set_heaviest_tipset(Arc::new(full_tipset.into_tipset()))?;
-                tracker.write().set_epoch(current_epoch);
-                metrics::LAST_VALIDATED_TIPSET_EPOCH.set(current_epoch);
+                let head = chainstore.heaviest_tipset();
+                // To avoid re-validating an unchanged head,
+                // which would fail when the head is from a snapshot that has only 900 most recent state roots.
+                if current_epoch != head.epoch() || full_tipset.key() != head.key() {
+                    let timer = metrics::TIPSET_PROCESSING_TIME.start_timer();
+                    validate_tipset(
+                        state_manager.clone(),
+                        &chainstore,
+                        bad_block_cache,
+                        full_tipset.clone(),
+                        genesis,
+                        invalid_block_strategy,
+                    )
+                    .await?;
+                    drop(timer);
+                    chainstore.set_heaviest_tipset(Arc::new(full_tipset.into_tipset()))?;
+                    tracker.write().set_epoch(current_epoch);
+                    metrics::LAST_VALIDATED_TIPSET_EPOCH.set(current_epoch);
+                }
```

Comment:
I don't understand this. Why do we fail if we validate HEAD?

Reply:
When we import a depth=900 snapshot at epoch 1900, it contains 900 state roots from epoch 1001 to 1900, which is just sufficient for validating epoch 1901, but if we re-validate epoch 1900, state root at epoch 1000 is required.

Reply:
Why do we re-validate epoch 1900?

Reply:
Okay, I think I see what is going on.

To validate tipset N, we execute tipset N-1 and check that the state-root matches the parent state-root of tipset N. So, when we import a snapshot with X epochs and receive tipset X+1, we'll execute the messages in tipset X. This causes Forest to fail because the tipsets in the snapshot must be assumed to be correct.

How does Lotus get around this?

Reply:
In any case, I'm fairly sure this constraint is incorrect: `current_epoch != head.epoch() || full_tipset.key() != head.key()`. We need a more principled solution.

Reply:
Another option could be populating the validation result cache for the snapshot head after import, what do you think?

Reply:
You're skipping the validation of the heaviest tipset from the snapshot. I don't think that is correct.

Reply:
> Another option could be populating the validation result cache for the snapshot head after import, what do you think?

We need the state-root hash to populate the validation result cache. We can't get it until we've received the next tipset.

Reply:
> You're skipping the validation of the heaviest tipset from the snapshot. I don't think that is correct.

That's how lotus works. When importing the same depth=900 snapshot, lotus succeeds only because it does not validate the snapshot head

Reply:
Sounds good.

Reply:
Fix removed. I will continue the convo in the issue thread and attach the relevant lotus code.